### PR TITLE
Upgrade to python 3.9.9

### PR DESCRIPTION
**Pull request by fbordignon**
I am issuing this PR to upgrade python to version 3.9.8

Two things that we need to keep in mind or discuss:

~~1 - There is a workaround on [line 539 of cmake/libpython/CMakeLists.txt](https://github.com/fbordignon/python-cmake-buildsystem/blob/7b3562e715872c72f198ae2f82d93795c7d2b2f7/cmake/libpython/CMakeLists.txt#L539) that copies the pyton39.lib to python3.lib because it was not being built anymore. I suspect it is because there is no more .def files in newer versions of python, so the target libpython3-shared is not generating python3.lib anymore. The only target that needs it is the extension_xxlimited which is a template of how to implement a module that declares objects, whatever that is. I believe it is not used by python because it is an example, but I am not sure.~~ Edited there isn't anymore

2 - ctypes depends on libffi, which is not distributed with python anymore on windows. Hence, if one needs ctypes one must provide the libraries and include paths `FFI_LIBRARIES` and `FFI_INCLUDE_DIRS`. I believe it is ok to leave it like that and the application developer can include the libffi on their build system if ctypes is needed.

### CODE REVIEW

**Comment by fbordignon** [comment]
This last commit bb29428 fixed the latest issues I was having with python3.dll on windows. I did not test on linux.

I don't know how to fix the circleci build, if someone can point me how to do it I can try, but I've never worked with it.

I've successfully integrated this branch into 3D Slicer and the [PR](https://github.com/Slicer/Slicer/pull/6026) is building and working on Windows.

Hence, I trust this PR is ready to be merged after the circleci fixes.

**Comment by jamesobutler** [comment]
> I don't know how to fix the circleci build, if someone can point me how to do it I can try, but I've never worked with it.

Likely that it needs to be using a CircleCI 2.0 configuration now which is in PR over at https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/276

**Comment by fbordignon** [comment]
Although the circle ci reports success, some tests are failing. I am trying to fix them but it will take some time.
The official windows build of python [includes a few libraries that are downloaded for building python](https://github.com/python/cpython-bin-deps) via this [script](https://github.com/python/cpython/blob/main/PCbuild/get_externals.bat). The tests expect some python modules that depend on them to be available. So I am thinking in including them on this build as well

```
bzip2-1.0.6
libffi-3.3.0
openssl-bin-1.1.1l
sqlite-3.35.5.0
tcltk-8.6.9.0
xz-5.2.2
zips
zlib-1.2.11/ 
```

Some of them are already included.

**Comment by jamesobutler** [comment]
Minor update to do: add support for Python 3.9.9 sources recently [released](https://pythoninsider.blogspot.com/2021/11/python-399-hotfix-release-is-now.html?m=1).

**Comment by fbordignon** [comment]
Ok, 13 tests failing on linux and 12 or so failing on windows. I don't know what to do anymore for them to pass.

There is an issue with the detection of readline that is bugging me
[line 378 of ConfigureChecks.cmake](https://github.com/fbordignon/python-cmake-buildsystem/blob/dbcdb1fed1c4e3ccfc4079e593470695f979fae9/cmake/ConfigureChecks.cmake#L378) goes:
```
check_include_files("stdio.h;readline/readline.h" HAVE_READLINE_READLINE_H)
```
But nothing is set on `HAVE_READLINE_READLINE_H`, whereas if I check `stdio.h` and `readline/readline.h` separately they are both found.

**Comment by fbordignon** [comment]
6 tests failing on Linux:
```
The following tests FAILED:
	 36 - test_c_locale_coercion (Failed)
	149 - test_gc (Failed)
	255 - test_posix (Failed)
	277 - test_regrtest (Failed)
	304 - test_socket (Failed)
	328 - test_symbol (Failed)
```
Other 6 on windows:
```
        test_cmd_line
	test_cmd_line_script
	test_ctypes	
	test_site
        test_venv
	test_zoneinfo
```

I don't know what to do to fix them. They seem very specific and not related to a missing module compilation.

**Comment by fbordignon** [comment]
test_ctypes is failling on windows because it assumes a dynamically linked sqlite3, which is not true. If we ignore [this test](https://github.com/python/cpython/blob/f4afc53bf68c8ded20b281cd1baa88a679b4a3fd/Lib/ctypes/test/test_loading.py#L121), it succeeds. Maybe others are also relative to the way we build the dependencies.

Does the previous versions of the builds passes all tests? Can we assume it is fine the way it is now?

**Comment by fbordignon** [comment]
Summary of latest changes:
- Updated to 3.9.9 as it is a minor change only patching a few files.
- Added the windows config for circleci. I've built for windows and linux successfully.
- Minor fix for posix threads definition

I've [Built](https://github.com/Slicer/Slicer/pull/6026#issuecomment-981592298) https://github.com/Slicer/Slicer/pull/6026/commits/7510d12428fcd273e165e04e4438208f19780a5b with this latest commit de81edb on windows and linux successfully.

**Comment by fbordignon** [comment]
@jcfr I am satisfied with the results of this PR and am not working on this actively as I've found no more significant issues for my use case (3D Slicer).
I am available to solve and/or discuss the issues that arises at the revision of this PR.

**Comment by jamesobutler** [comment]
@jcfr Do you have time to provide review here to help finalize the integration of this work?

**Comment by jcfr** [comment]
Thanks again @fbordignon, @dbrnz, @dand-oss  and everyone who contributed code as well as time for reviewing and improving this pull request. This is outstanding 💯 🙏 

This week, I plan to review and finalize the integration.

Thanks again for your patience

**Comment by fbordignon** [comment]
@jcfr I did not care to keep the compatibility with previous versions of python. I can see that it may be desirable but we can try to keep a branch for each major version that would compile with all the minor versions i.e. branch 3.9 will compile 3.9.1 3.9.2 and so on.
If that is not an option, I can try to work on compatibility with 3.7+

**Comment by jcfr** [comment]
Hi @fbordignon and happy new year,

I plan to spend more time in the next few days. After that, I suggest re re asses where we are at.

File: readme.txt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+git checkout master
+git rebase -i ci-376
+git push -f
+
+ci-376 - been pushing this to CI
+   https://ci.appveyor.com/project/dand-oss/python-cmake-buildsystem
+   https://travis-ci.org/dand-oss/python-cmake-buildsystem
+
+upgrade-py376 - pull request to upstream
```
Comment (jamesobutler) [inline readme.txt:27]: This just a dev notes txt file that isn’t to be merged into the final commit?

**Comment by jcfr** [comment]
This pull-request is superseded by https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/292 and it will be closed once #292 has been integrated.

**Pull request closed without merging**